### PR TITLE
fix infinity loop

### LINKED ISSUE #408: where(id: Array(Int)) generates infinite loop
example:
```
      it "errors queries with ids" do
        review1 = Review.create(name: "one")
        review2 = Review.create(name: "two")

	found = Review.where(id: [review1.id, review2.id]).select
        found[0].id.should eq review2.id
        found[1].id.should eq review1.id
      end
```
```
[0x55cd10fa4639] *Granite::Query::Builder(Review) +169
[0x55cd10fa4495] *Granite::Query::Builder(Review) +549
[0x55cd10fa4266] *Granite::Query::Builder(Review) +134
[0x55cd10fa41ce] *Granite::Query::Builder(Review) +78
[0x55cd10fa4171] *Granite::Query::Builder(Review) +49
[0x55cd10da204d] *Review +45
[0x55cd10b4abbc] ~procProc(Nil) +284
[0x55cd10d64257] *Spec::Example#internal_run<Time::Span, Proc(Nil)>:(Array(Spec::Result) | Nil) +231
[0x55cd10d64137] *Spec::Example#run:Nil +1143
[0x55cd10d63319] *Spec::ExampleGroup#internal_run:Nil +137
[0x55cd10d631c3] *Spec::ExampleGroup#run:Nil +419
[0x55cd10d34f07] *Spec::RootContext#run:Nil +167
[0x55cd10b0de75] ~procProc(Int32, (Exception | Nil), Nil) +101
[0x55cd10c452e3] *Crystal::AtExitHandlers::run<Int32, (Exception+ | Nil)>:Int32 +195
[0x55cd10fed00b] *Crystal::main<Int32, Pointer(Pointer(UInt8))>:Int32 +139
[0x55cd10afbb96] main +6
[0x7fef2368fb97] __libc_start_main +231
[0x55cd10aeb1ca] _start +42
[0x0] ???
```

### PR DESCRIPTION
**Pull request by OpakAlex**
Fixes #408

### CODE REVIEW

File: src/granite/query/builder.cr [hunk 1]
```diff
@@ -75,19 +75,13 @@ class Granite::Query::Builder(Model)
...
-      and(field: field.to_s, operator: :eq, value: value)
-    end
+  def and(field : (Symbol | String), operator : Symbol, value : Granite::Columns::Type)
+    @where_fields << {join: :and, field: field.to_s, operator: operator, value: value}
 
     self
   end
 
-  def and(field : (Symbol | String), operator : Symbol, value : Granite::Columns::Type)
+  def and(field : (Symbol | String), operator : Symbol, value : Array(String | Int32 | Int64))
```
Comment (Blacksmoke16) [inline src/granite/query/builder.cr:84]: I still don't think this should be required.

Comment (OpakAlex) [reply src/granite/query/builder.cr]: you are right! Removed it

**Comment by OpakAlex** [comment]
@Blacksmoke16 fixed

**Comment by OpakAlex** [comment]
@drujensen can you review too?

File: spec/granite/querying/query_builder_spec.cr [hunk 2]
```diff
@@ -3,6 +3,22 @@ require "../../spec_helper"
 describe Granite::Query::BuilderMethods do
   describe "#where" do
     describe "with array argument" do
+      it "errors queries ids fields" do
```
Comment (drujensen) [inline spec/granite/querying/query_builder_spec.cr:6]: description of this test is confusing.  consider `it "correctly queries all rows with a list of id values" do`.

File: spec/granite/querying/query_builder_spec.cr [hunk 3]
```diff
@@ -3,6 +3,22 @@ require "../../spec_helper"
...
+      end
+
+      it "correctly queries with ids fields which doest exists" do
+        Review.create(name: "one")
+        Review.create(name: "two")
+
+        found = Review.where(id: [1001, 1002]).select
+        found.size.should eq 0
+      end
       it "correctly queries string fields" do
```
Comment (drujensen) [inline spec/granite/querying/query_builder_spec.cr:22]: Add a test for `and` and `or` operators.

File: src/granite/query/builder.cr [hunk 4]
```diff
@@ -99,15 +87,32 @@ class Granite::Query::Builder(Model)
...
   end
 
+  def and(**matches)
+    and(matches)
+  end
+
+  def and(matches)
+    value = matches[:value]
+    if value.is_a?(Array)
+      and(matches[:field].to_s, matches[:operator], value.compact)
```
Comment (drujensen) [inline src/granite/query/builder.cr:97]: Shouldn't the default operator be `:in` when it's type is an Array?
```
and(matches[:field].to_s, matches[:operator] || :in, value.compact)
```

File: src/granite/query/builder.cr [hunk 5]
```diff
@@ -99,15 +87,32 @@ class Granite::Query::Builder(Model)
...
   def or(**matches)
     or(matches)
   end
 
   def or(matches)
-    matches.each do |field, value|
-      or(field: field.to_s, operator: :eq, value: value)
+    value = matches[:value]
+    if value.is_a?(Array)
+      or(matches[:field].to_s, matches[:operator], value.compact)
```
Comment (drujensen) [inline src/granite/query/builder.cr:112]: Same here.

File: spec/granite/querying/query_builder_spec.cr (hunk 2 above)
Comment (OpakAlex) [reply spec/granite/querying/query_builder_spec.cr]: changed

File: src/granite/query/builder.cr (hunk 4 above)
Comment (OpakAlex) [reply src/granite/query/builder.cr]: i added spec.

File: spec/granite/querying/query_builder_spec.cr (hunk 3 above)
Comment (OpakAlex) [reply spec/granite/querying/query_builder_spec.cr:44]: done

**Comment by OpakAlex** [comment]
@drujensen fixed your comments. Can you recheck?

**Pull request merged**